### ISSUE TITLE
Is it possible to restart the dask client as and when within a code?

### ISSUE THREAD

**Issue by justlearn-spec** [dask/distributed#3605]
I want to see if there is any performance improvement if I keep modifying n_workers and threads_per_client as and when needed.

import dask.dataframe as dd
from dask.distributed import Client, LocalCluster
cluster = LocalCluster(n_workers=8, threads_per_worker=5)
client = Client(cluster)
... (E.g., Finding fit and predict)
cluster = LocalCluster(n_workers=5, threads_per_worker=8) 
client = Client(cluster)
... (E.g., Performing some computations)

I attempted with the client.restart() and client.shutdown() then imported new combinations in cluster variable, but it vain.

In fact, client.restart() has reset n_workers and threads_per_worker.

**Comment by mrocklin** [comment]
Hi @justlearn-spec , I'm not sure I understand your question.  

The code snippet that you include doesn't use restart or shutdown.  Can you include a minimal example of what you were trying to do and what you expected?

**Comment by justlearn-spec** [comment]
Hi @mrocklin,

I'm trying to tune one large snippet which does the below,

1. Import CSV of 10GB into Dask Dataframe. (At the beginning n_workers set to 8 and threads_per_worker set to 5)
2. Drop some column and add Target Column in the dataframe
3. Label Encoding
4. Finding a fit
5. Predict
6. Calculations using NumPy/Dask array (da.nansum/da.nanmean)

Server spec: 
RAM: 252 GB
Cores: 40

Here, the calculations using Numpy (OR) Dask array take a long time which I wanted to tune. So, I thought of modifying n_workers and threads_per_worker to some other value before the 6th step and measure the performance. 

In a nutshell, if I execute the below code,
from dask.distributed import Client, LocalCluster
cluster = LocalCluster(n_workers=8, threads_per_worker=5)
client = Client(cluster)
cluster = LocalCluster(n_workers=10, threads_per_worker=4)
client = Client(cluster)

I want to see n_workers as 10 and threads_per_worker as 4

**Comment by mrocklin** [comment]
So I think that your question is 

> "if I have two active clusters with a client attached to each, which is used by default?"

If so, the answer is ...

> The most recently created Client is used by default

That may not be what you are asking though.  I apologize if I have misunderstood.

**Comment by justlearn-spec** [comment]
Your understanding is correct @mrocklin . The most recently created Client is not reflecting.

![image](https://user-images.githubusercontent.com/61821667/77246619-3d025280-6c4f-11ea-9dde-268460b63cc6.png)

**Comment by mrocklin** [comment]
What are you trying to show in that example?  When you call `LocalCluster` you are making a new LocalCluster, not referring to the most recently made one.

**Comment by justlearn-spec** [comment]
I'm very new to this. 
Sorry. That was my bad and this is a usage question.  I thought LocalCluster (running twice) will modify the existing one. 

Should I execute the below code to activate n_workers=5 and threads_per_worker=8 on the same port where n_workers was set to 8 and threads_per_worker=5?

from dask.distributed import Client
client = Client(n_workers=8, threads_per_worker=5)
client
client = Client(n_workers=5, threads_per_worker=8)
client

**Comment by mrocklin** [comment]
You are creating two Dask local clusters with different configurations.
You are not changing an existing one.  This is fine, the old one will be
cleaned up from regular Python GC.

**Comment by justlearn-spec** [comment]
Thanks for your response @mrocklin 

How can I override the existing client? I tried with client.remove_worker (taken from previous posts), but it didn't help.

**Comment by justlearn-spec** [comment]
The reason why I wanted to modify the existing client is, below code is running for 3 hours and I want to check converting threads from processes (OR) modifying the n_workers expedites the job processing or not. 
================================================
Prediction was done using Dask-LightGBM (220 million records)
Preprocessing: Train test split
y_preds_lgbm_train -> est.predict(X_train) 
================================================
def eval_results(yhat_mean, y):
	calc = dd.DataFrame.abs(y - yhat_mean)
	mad = da.nanmean(calc).compute()
	print('MAD: {}'.format(mad))
	mse = da.nanmean(da.square(y - yhat_mean)).compute()
	print('MSE: {}'.format(mse))
	mape = np.nansum(calc) / np.nansum(y)
	print('MAPE: {}'.format(mape))
	smape = 100. * da.nanmean(calc / ((np.abs(y) + np.abs(yhat_mean)) / 2.)).compute()
	print('SMAPE: {}'.format(smape))
	md = da.nanmean(y - yhat_mean).compute()
	print('MD: {}'.format(md))
	mean_y = da.nanmean(y).compute()
	print('mean(y): {}'.format(mean_y))
	mean_yhat = da.nanmean(yhat_mean).compute()
	print('mean(y_hat): {}'.format(mean_yhat))

print(eval_results(y_preds_lgbm_train, y_train))

**Comment by mrocklin** [comment]
Then you can create a new process with a different LocalCluster.  Swapping out client/cluster during a computation is atypical.

If you're looking for help on understanding performance on a particular workload then I recommend reading through https://docs.dask.org/en/latest/understanding-performance.html

This issue has turned into an extended Q&A, so I'm going to close.  If you come across a specific bug or performance issue then I encourage you to raise another issue with a minimal reproducible example.  (Please read through http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports)
